### PR TITLE
build size-check enhancement

### PR DESCRIPTION
**Pull request by mtei**
## Description

Changed to display a warning when the free size of compilation result is less than 512 bytes. :-)

Example:
```
$ make crkbd:default
.....
Copying crkbd_rev1_default.hex to qmk_firmware folder                        [OK]
Checking file size of crkbd_rev1_default.hex                                 [WARNINGS]

 * The firmware size is the limit soon - 28388/28672 (284 bytes free)
```

```
$ make crkbd:all
QMK Firmware 0.6.321
Making crkbd/rev1 with keymap default                                        [WARNINGS]

 * The firmware size is the limit soon - 28388/28672 (284 bytes free)
Making crkbd/rev1 with keymap drashna                                        [OK]
Making crkbd/rev1 with keymap edvorakjp                                      [OK]
Making crkbd/rev1 with keymap jarred                                         [OK]
Making crkbd/rev1 with keymap like_jis                                       [OK]
Making crkbd/rev1 with keymap omgvee                                         [OK]
Making crkbd/rev1 with keymap thefrey                                        [WARNINGS]

 * The firmware size is the limit soon - 28400/28672 (272 bytes free)
```

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: message.mk [hunk 1]
```diff
@@ -80,3 +80,4 @@ MSG_CHECK_FILESIZE = Checking file size of $(TARGET).hex
 MSG_FILE_TOO_BIG = $(ERROR_COLOR)The firmware is too large!$(NO_COLOR) $(CURRENT_SIZE)/$(MAX_SIZE) ($(OVER_SIZE) bytes over)\n
 MSG_FILE_TOO_SMALL = The firmware is too small! $(CURRENT_SIZE)/$(MAX_SIZE)\n
 MSG_FILE_JUST_RIGHT = The firmware size is fine - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
+MSG_FILE_NEAR_LIMIT = The firmware size is the limit soon - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
```
Comment (noroadsleft) [inline message.mk]:
```suggestion
MSG_FILE_NEAR_LIMIT = The firmware size is nearly too large - $(CURRENT_SIZE)/$(MAX_SIZE) (only $(FREE_SIZE) bytes free)\n
```

Comment (mtei) [reply message.mk]: Thank you.

**Comment by fauxpark** [comment]
Considering how wildly the filesize varies depending on the enabled features, I don't know how useful this will be. It doesn't seem like there is that fine of control over it.

Funnily enough [I also have a Gist](https://gist.github.com/fauxpark/de69f3a016c94663cfbd2b7d6581de92) that looks pretty much identical to this, but with a threshold of 1024 bytes. I didn't get around to submitting it due to the above.

**Comment by mtei** [comment]
By looking at the test results of the existing PR, I can know which keyboard will reach its limit soon. This is useful information when changing core modules. I can try a build by choosing a keyboard that is likely to exceed the limit.

**Comment by XScorpion2** [comment]
Very nice!

**Comment by drashna** [comment]
@mtei I agree that this is very nice.  It gives us a good look at what boards are close to the limit.  Also, it may help to start get some optimization done.

**Comment by fauxpark** [comment]
In my testing I had difficulty getting builds which were within the range of this warning, even at 1024 bytes, by just juggling rules.mk features. They were almost always under or oversize. What sort of changes is this intended to warn against? Layers?

**Comment by drashna** [comment]
Several.  For one, Travis CI uses avr gcc 5.4.0, which compiles larger than newer versions.  So it may flag firmware.  Also, it may help to identify feature creep, in ... when boards have a high numbered layer, wasting space. 

Also, it feels like a good move for users, as it makes it obvious when you're starting to run low on firmware space.  (like in my case).


For a majority of boards, this won't matter.  But there are a few boards .... where this would. And a number of those are already showing up in the Travis CI check here

**Comment by fauxpark** [comment]
> Travis CI uses avr gcc 5.4.0, which compiles larger than newer versions. So it may flag firmware.

That just sounds like a reason *not* to implement this, tbh. Wouldn't it be a false positive compared to when you build it yourself?

> it makes it obvious when you're starting to run low on firmware space

Maybe, if enabling certain features didn't just push it right over the limit. See some of the comments in the build options - these might be outdated numbers but enabling mouse keys for example apparently adds around 4700 bytes, and turning on media keys adds another ~450 (I guess media keys are usually enabled anyway though).

The nonconsecutive layers thing I agree with, but I don't know how common that is.

**Comment by drashna** [comment]
> That just sounds like a reason _not_ to implement this, tbh. Wouldn't it be a false positive compared to when you build it yourself?

MSYS uses the same version of AVR GCC, actually.  So, it wouldn't necessarily be a false positive. 

And there may be distros using older versions, as well.   So, using the oldest version is more reasonable than a newer version that not everyone has.  That way, users don't have to worry about "why it won't compile the default keymap".    


>The nonconsecutive layers thing I agree with, but I don't know how common that is.

It's pretty common. A lot of the boards using the tri layer stuff use it, due to copy-pasta

> Maybe, if enabling certain features didn't just push it right over the limit. See some of the comments in the build options - these might be outdated numbers but enabling mouse keys for example apparently adds around 4700 bytes, and turning on media keys adds another ~450 (I guess media keys are usually enabled anyway though).

The numbers are worse than outdated.  They're completely compiler version dependant.  And most of them are not accurate, and haven't been for a long while. 

The other issue, is that we need to have "reasonable defaults" for the Configurator, too.  We can't just turn off everything. At least until the Configurator has the ability to toggle abilities. 


Either way, I can definitely see reasons for both adding these and reasons not to.  I personally do like it, but I'll defer to others, here.

**Comment by noroadsleft** [comment]
I tested this PR during my review earlier. I don't know how @drashna did it, but my MSYS avr-gcc is even older than 5.4.0 (I'm running Windows 8.1):

```
$ avr-gcc --version
avr-gcc.exe (AVR_8_bit_GNU_Toolchain_3.5.4_1709) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

For me, the crkbd on current master barely fits the available space:

```
$ qmk crkbd:default
QMK Firmware 0.6.322
Making crkbd/rev1 with keymap default

avr-gcc.exe (AVR_8_bit_GNU_Toolchain_3.5.4_1709) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

...

Copying crkbd_rev1_default.hex to qmk_firmware folder                                               [OK]
Checking file size of crkbd_rev1_default.hex                                                        [OK]
 * The firmware size is fine - 28612/28672 (60 bytes free)
```

**Comment by fauxpark** [comment]
@noroadsleft the AVR toolchain provided by Microchip/Atmel was "updated" sometime recently I think.

**Comment by drashna** [comment]
@noroadsleft Yeah, it was updated, and a while back, I updated the MSYS script to grab that version instead. If you re-run the script, it should update the components.

**Comment by noroadsleft** [comment]
@drashna and re-running the install script just broke my everything:

```
$ qmk crkbd/rev1:default
QMK Firmware 0.6.322
Making crkbd/rev1 with keymap default

/bin/sh: avr-gcc: command not found
sh: avr-gcc: command not found
make[1]: *** [tmk_core/rules.mk:253: gccversion] Error 127
make[1]: *** Waiting for unfinished jobs....
sh: avr-gcc: command not found
sh: avr-gcc: command not found
make[1]: *** [tmk_core/rules.mk:369: .build/obj_crkbd_rev1_default/compiler.txt] Error 127
Make finished with errors
make: *** [Makefile:551: crkbd/rev1:default] Error 1
```

**Comment by mtei** [comment]
Changed default margin 512 to 1024.

```
$ make crkbd:all
QMK Firmware 0.6.321
Making crkbd/rev1 with keymap default                                     [WARNINGS]

 * The firmware size is nearly too large - 28388/28672 (284 bytes free)
Making crkbd/rev1 with keymap drashna                                     [WARNINGS]

 * The firmware size is nearly too large - 28038/28672 (634 bytes free)
Making crkbd/rev1 with keymap edvorakjp                                   [OK]
Making crkbd/rev1 with keymap jarred                                      [WARNINGS]

 * The firmware size is nearly too large - 27974/28672 (698 bytes free)
Making crkbd/rev1 with keymap like_jis                                    [OK]
Making crkbd/rev1 with keymap omgvee                                      [OK]
Making crkbd/rev1 with keymap thefrey                                     [WARNINGS]

 * The firmware size is nearly too large - 28400/28672 (272 bytes free)
```

Added SIZE_MARGIN variable.

```
$ make SIZE_MARGIN=275 crkbd:all
QMK Firmware 0.6.321
Making crkbd/rev1 with keymap default                                     [OK]
Making crkbd/rev1 with keymap drashna                                     [OK]
Making crkbd/rev1 with keymap edvorakjp                                   [OK]
Making crkbd/rev1 with keymap jarred                                      [OK]
Making crkbd/rev1 with keymap like_jis                                    [OK]
Making crkbd/rev1 with keymap omgvee                                      [OK]
Making crkbd/rev1 with keymap thefrey                                     [WARNINGS]

 * The firmware size is nearly too large - 28400/28672 (272 bytes free)
```

File: message.mk [hunk 2]
```diff
@@ -80,3 +80,4 @@ MSG_CHECK_FILESIZE = Checking file size of $(TARGET).hex
 MSG_FILE_TOO_BIG = $(ERROR_COLOR)The firmware is too large!$(NO_COLOR) $(CURRENT_SIZE)/$(MAX_SIZE) ($(OVER_SIZE) bytes over)\n
 MSG_FILE_TOO_SMALL = The firmware is too small! $(CURRENT_SIZE)/$(MAX_SIZE)\n
 MSG_FILE_JUST_RIGHT = The firmware size is fine - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
+MSG_FILE_NEAR_LIMIT = The firmware size is nearly too large - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
```
Comment (drashna) [inline message.mk]: I'm still not happy with the wording. Maybe @mechmerlin has a better suggestion, but I think this sounds better: 

```suggestion
MSG_FILE_NEAR_LIMIT = The firmware size is close to the controller's max capacity - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
```

Comment (fauxpark) [reply message.mk]: I'm rather partial to the wording I used: `The firmware size is approaching the maximum`

Comment (drashna) [reply message.mk]:
```suggestion
MSG_FILE_NEAR_LIMIT = The firmware size is approaching the maximum - $(CURRENT_SIZE)/$(MAX_SIZE) ($(FREE_SIZE) bytes free)\n
```

**Comment by drashna** [comment]
Awesome, and thanks for all the work that you've done for QMK Firmware @mtei!!

**Pull request merged**